specifications/mempool/README.md

```diff
@@ -0,0 +1,280 @@
+# Mempool specification
+
+## Abstract
+
+This document describes the spec of a Libra node's mempool component,
+which is responsible for accepting transaction submitted from other nodes
+or directly to the node and relaying them to consensus.
+
+This document is organized as follows:
+
+1. Overview: high-level outline of the mempool architecture
+2. Module Specifications: spec of mempool components
+3. Network Protocol: the Mempool LibraNet protocol used to participate in the Libra
+Payment Network
+4. Abstracted Modules: external components that mempool interfaces with
+
+## Overview
+
+Mempool is a system component of the Libra node that is responsible for accepting, 
+validating, and prioritizing, and broadcasting pending transactions. 
+
+Mempool consists of a local memory-buffer that holds the transactions that 
+are waiting to be executed (core mempool) and a network stack (shared mempool) 
+that manages communication with other peers, as well as other system components.
+
+The [JSON RPC](../../json-rpc/README.md) module proxies transactions to mempool.
+Mempool validates transactions and stores it locally.
+
+On the validator node, [consensus](../consensus/README.md) pulls ordered transactions
+from mempool to form a block. 
+
+Mempool is notified once a transaction is fully executed and committed. Mempool then
+drops this transaction from its local state. 
+
+A transaction remains in mempool at least until the earliest of the following:
+* committed
+* client-specified expiration time of the transaction is reached
+* mempool-defined TTL is reached, so that no transaction stays in mempool for longer
+than this TTL
+
+
+## Module Specifications
+
+This section (1) outlines the modules of mempool and their core 
+functionalities, and (2) defines some mempool terminology.
+Note: the function signatures defined for each component's functionality 
+is part of a _sample_ interface not strictly required by the specification.
+
+
+### Terminology
+
+#### Transaction states in Mempool
+
+A transaction in mempool can be (but not limited to):
+
+* validated: checked that the following conditions are met
+
+  * the sequence number of the transaction is not older than the latest transaction in the blockchain for the same account
+  * transaction has enough gas
+  * transaction signature is valid
+  * transaction has not expired at the time of validation
+  * transaction payload is a valid Move script
+
+* consensus-ready: transaction has some probability of being included in the next block
+
+  * transaction is validated
+  * the sequence number of the transaction is the next sequence number of the sender account, or it is sequential to it
+
+* broadcast-ready: transaction is eligible to be broadcasted to a peer node
+
+  * transaction is consensus-ready
+  * transactions that a validator receives from another validator peer are not broadcast-ready, even if they are consensus-ready
```

Comment:
probably more readable if you just wrote this out:

transaction is eligible to be broadcasted to a peer node if it is consensus-ready; **however**, transactions received from another validator peer (or peer on the same network) will **never** be broadcast ready even if they are consensus-ready.